### PR TITLE
[ci] Skip unit tests requiring git if git is not available

### LINKED ISSUE #1582: Skip unit tests using `git` when `git` is not available

### PR DESCRIPTION
**Pull request by teojgo**
Fixes #1582

### CODE REVIEW

**Review by ekouts — APPROVED** [review]

File: unittests/test_utility.py [hunk 1]
```diff
@@ -235,7 +235,20 @@ def test_is_url():
     assert not osext.is_url(repo_ssh)
 
 
-def test_git_repo_hash(monkeypatch):
+@pytest.fixture
+def no_git_installed():
```
Comment (vkarak) [inline unittests/test_utility.py:239]: I would inverse the check and call the fixture `git_only`.

File: unittests/test_utility.py [hunk 2]
```diff
@@ -235,7 +235,20 @@ def test_is_url():
...
+    try:
+        completed = osext.run_command(f'git --version', check=True, log=False)
```
Comment (vkarak) [inline unittests/test_utility.py:241]:
```suggestion
        completed = osext.run_command('git --version', check=True, log=False)
```

File: unittests/test_utility.py [hunk 3]
```diff
@@ -235,7 +235,20 @@ def test_is_url():
...
+    except (SpawnedProcessError, FileNotFoundError):
+        return True
+
+    return False
+
+
+def test_git_repo_hash(monkeypatch, no_git_installed):
+    if no_git_installed:
+        pytest.skip('no git installation found on system')
```
Comment (vkarak) [inline unittests/test_utility.py:250]: The skipping can happen in the fixture.

**Review by vkarak — CHANGES_REQUESTED** [review]

File: unittests/test_utility.py (hunk 1 above)
Comment (vkarak) [reply unittests/test_utility.py:239]: If you skip the test inside the fixture, you don't need to return anything.

**Comment by vkarak** [comment]
Can you also update the Dockerfiles for the modules systems?

File: ci-scripts/dockerfiles/Lmod.dockerfile [hunk 4]
```diff
@@ -7,7 +7,9 @@ FROM ubuntu:20.04
 ENV TZ=Europe/Zurich
 ENV DEBIAN_FRONTEND=noninteractive
 ENV _LMOD_VER=8.4.12
-WORKDIR /root
+
+# ReFrame user
```
Comment (vkarak) [inline ci-scripts/dockerfiles/Lmod.dockerfile:11]: Can you expand the comment there why we need a special user? If it's a standard practice, there is no need.

File: unittests/test_utility.py [hunk 5]
```diff
@@ -236,18 +236,18 @@ def test_is_url():
...
 
 
-def test_git_repo_hash(monkeypatch, no_git_installed):
-    if no_git_installed:
-        pytest.skip('no git installation found on system')
+def test_git_repo_hash(monkeypatch, git_only):
+    try:
+        completed = osext.run_command('git status', check=True, log=False)
+    except (SpawnedProcessError, FileNotFoundError):
```
Comment (vkarak) [inline unittests/test_utility.py:249]: Why don't you simply check if `.git` exists?

**Review by vkarak — CHANGES_REQUESTED** [review]

File: ci-scripts/dockerfiles/Lmod.dockerfile (hunk 4 above)
Comment (teojgo) [reply ci-scripts/dockerfiles/Lmod.dockerfile:11]: It's just better to have a normal user without root privileges to be as close as possible to the typical use case.

File: unittests/test_utility.py (hunk 5 above)
Comment (teojgo) [reply unittests/test_utility.py:249]: I think it's better to rely on git to check if it's a valid repo.

Comment (vkarak) [reply unittests/test_utility.py:249]: My only problem is whether this is worth it. This four lines of code and much more verbose instead of just two.

Comment (teojgo) [reply unittests/test_utility.py:249]: `.git` can exist but in case it's not a proper repo you cannot get the hash.

Comment (vkarak) [reply unittests/test_utility.py:249]: Ok, I see the scenario that sb downloads the source (without cloning), has git and runs the unit tests, so this check has a meaning. I'm not gonna argue further about the way for writing it.

File: unittests/test_utility.py [hunk 6]
```diff
@@ -247,7 +260,7 @@ def test_git_repo_hash(monkeypatch):
     assert osext.git_repo_hash() is None
```
Comment (vkarak) [inline unittests/test_utility.py:260]: I think here we should add a case that monkeypatches the `GIT_DIR`, so as to check the case that the `rev-parse` returns an error.

**Review by vkarak — CHANGES_REQUESTED** [review]

**Review by vkarak — APPROVED** [review]

**Pull request merged**